File: src/Lessons/4-CountingElements/4.3-MaxCounters.go
package main

import "fmt"

func main() {

	fmt.Println(solution(5, []int{3, 4, 4, 6, 1, 4, 4}))
	//expect  [3, 2, 2, 4, 2]
}

func solution(numOfCounters int, someArray []int) []int {
	counters := make([]int, numOfCounters)
	counterMax := 0
	counterMaxTracker := 0

	//Loop through array
	for index := 0; index < len(someArray); index++ {
		num := someArray[index]
		counterIndex := num - 1

		if num >= 1 && num <= numOfCounters {
			//increase counter num by 1
			if counters[counterIndex] < counterMax {
				counters[counterIndex] = counterMax + 1
			} else {
				counters[counterIndex] += 1
			}
			if counters[counterIndex] > counterMaxTracker {
				counterMaxTracker = counters[counterIndex]
			}
		}
		if num == numOfCounters+1 {
			//max all counters by storing in counterMax
			counterMax = counterMaxTracker
		}
	}
	//Loop through our counters
	for index := 0; index < len(counters); index++ {
		counter := counters[index]
		//Max all counters less than the required min of countermax
		if counter < counterMax {
			counters[index] = counterMax
		}
	}

	return counters

}

File: src/Lessons/4-CountingElements/4.3-MaxCounters.js
function solution(N, A) {
    //Create our counters
    let array = [...A];
    let counterMax = 0;
    let counterMaxTracker = counterMax
    let counters = new Array(N);
    //Loop 1;
    for (let index = 0; index < array.length; index++) {
        const X = array[index];
        const num = X - 1;
        //if true increase
        if (X <= N && X >= 1) {
            //decrease by 1 to match our counter index
            if(!counters[num]) //if empty
                counters[num] = counterMaxTracker;

            if (counters[num] < counterMaxTracker) //if less than the current max value
                counters[num] = counterMaxTracker + 1;
            else
                counters[num] += 1;
            //store our current max value
            if (counters[num] > counterMax) counterMax = counters[num];
        }
        //if true max all counters with max value
        if (X == N + 1)
            counterMaxTracker = counterMax;
    }
    //Loop 2
    //Now loop to update all values below our most recent max value
    for (let index = 0; index < counters.length; index++) {
        if(!counters[index]) //if empty
            counters[index] = counterMaxTracker;
    
        else if(counters[index] < counterMaxTracker)
            counters[index] = counterMaxTracker; 
    }
    return counters;
}

const testArray = [3, 4, 4, 6, 1, 4, 4];
console.log(solution(5, testArray))
//expect 7